Classifier: add image placeholder for Single Image Viewer (Fix 2221)

shaunanoordin opened the pull request:
## PR Overview

packages: code change in `lib-classifier`, but can only be tested on `app-project`
Fixes #2221 

This PR adds an image placeholder for the Single Image Viewer, when the SIV is in the process of loading an image file for a Subject.

- The image placeholder ensures that the Single Image Viewer properly renders when the Subject resource loads, but before the image file is loaded.
  - Without this PR, the SIV only renders when both the Subject resource AND image file load.
  - The reason it needs to "properly render" at this stage is because it needs to create the _child <svg>_ element that the _parent container_ needs a _ref_ to. (See Analysis for deets. This is a trippy problem.) 
- This is a solution to the issue #2221, which can be more generally described as _"if your Classifier supports Drawing Tasks, and a user navigates away from the Classifier page then returns to the Classifier page, the Drawing Tasks will get borked."_
- ⚠️ Other Subject Viewers may be affected by this bug, but my focus is only on the SIV right now. Will need to check the others later (e.g. Video Viewer. Tagging @ErikOstlund as an FYI.)

### Analysis

**The Problem**

Do the following:

1. Load up `app-project` and find any Project with a Drawing Task, e.g. [I Fancy Cats](http://localhost:3000/projects/brooke/i-fancy-cats/classify/workflow/693?env=staging)
2. Open the Classifier page.
3. Make some annotations, esp. some drawing annotations
4. Without submitting, navigate to the Project home page. (click on the project navi links)
5. Return to the Classifier page. (click on the project navi links)
6. You'll note that the drawing tools/tasks on the Classifier has completely borked. Points/lines/etc won't render properly, and a bunch of error messages will be in the console.
7. Changing the Subject Viewer settings - e.g. zooming - sort of resets the drawing tools/task

**Why Is This Breaking?**

In short: combining _conditional rendering_ with _referenced child components_ causes things to break if even the _slightest_ thing is out of the strictly expected sequence.

**How Is This Breaking?**

Full story? Follow me on a journey.

In Step 2 above, User visits Classifier for the first time. The Subject Viewer (SingleImageViewerContainer, actually) will render 3 times. Here's what happens in each instance of SingleImageViewerContainer.render():

First visit, render 1:
> loadingState = loading
> img.src && img.naturalWidth undefined, so render null
> Fetch Image via onLoad()
> context.svg = nothing because no <svg> has been rendered.

First visit, render 2:
> loadingState = loading (still?)
> Fetch Image success!
> img.src && img.naturalWidth now defined, so render components!
> context.svg = nothing (still!) because no <svg> is only rendered after this is set.

First visit, render 3:
> loadingState = success (this is what triggers the 3rd render)
> context.svg = defined! Because the <svg> was rendered during the previous step

Okay, so everything is fine up to this point. But what happens when you go to the project home page, then revisit the Classifier page?

Second visit, render 1:
> loadingState = success already ❗ 
> img.src && img.naturalWidth is undefined, (despite loadingState is sucess!) so render null
> Fetch Image via onLoad()
> context.svg = nothing because no <svg> has been rendered.

Second visit, render 2:
> loadingState = success already
> Fetch Image success!
> img.src && img.naturalWidth now defined, so render components!
> context.svg = nothing (still!) because no <svg> is only rendered after this is set.

Because loadingState = success during the second visit to the Classifier, the third render isn't triggering. This means that the reference for context.svg is never set, which causes loads of problems.

**What's the solution?**

Simple: placeholder images.

See, I think the answer here is to remove the _conditional rendering_ aspect, which is royally screwing up the necessary parent-child reference passing. The SingleImageViewerContainer should always render all its children (therefore ensuring component references are consistent). And if the SIVC is always rendering its children, it needs to show something where the Image file hasn't fully loaded... hence, show a placeholder image while the Subject's Image file is still being processed.

### Status

WIP

TODO:
- [x] Add a placeholder to fix bug
- [x] Follow [Invision design](https://projects.invisionapp.com/d/main#/console/12923997/442007763/preview)
- [x] Fix tests

eatyourgreens commented:
There's a design for a subject placeholder in the classifier. It's come up a couple of times for Engaging Crowds, so it would be good to get it implemented. https://github.com/zooniverse/front-end-monorepo/pull/1966#pullrequestreview-565501715

shaunanoordin commented:
Oh that is brilliant! Thanks Jim! 👍 

Surfacing the deets for easy reference:
- [Invision design 🔗 ](https://projects.invisionapp.com/d/main#/console/12923997/442007763/preview)
- [Slack convo 🔗](https://zooniverse.slack.com/archives/CTXM9AZL6/p1610380489094700?thread_ts=1610373562.089700&cid=CTXM9AZL6)

shaunanoordin commented:
## PR Update

OK, this is MVP functional! Here are the changes required to fix the issue of _"Drawing Tasks break when RE-visiting Classifier page."_

**SubjectImageViewerContainer:**
- The SubjectImageViewerContainer should always render _as long as the Subject resource hasn't errored._
- If the image file hasn't loaded, prepare placeholder values for the image. i.e. set _src, naturalHeight, and naturalWidth_ so children elements don't crash altogether.
  - ⚠️ As of the latest commit, this is literally a _placeholder_ placeholder image. The "correct" placeholder image should be the Zooniverse logo provided by Becky in the Invision design.
- Remove the "has the image file loaded?" clause for conditional rendering, i.e. remove `if (src && naturalWidth)` in the render() code.
  - We MUST always render the sub-components of SubjectImageViewerContainer, because otherwise there's a chicken-and-egg scenario where a parent requires a ref to a child so it can render another child but it can't render until the parent renders etc etc etc.

**SVGPanZoom:**
- The imageScale() function should always return a number, never a NaN. Default value is 1.0.
  - Strictly speaking, this isn't required to fix this issue (the NaN numbers eventually sort themselves out after a few render cycles), but it will prevent hidden console errors when trying to render Drawing Marks.

Tests are still borked, but that's a later problem.

shaunanoordin commented:
@snblickhan I'm tagging you in this PR so you can keep an eye on progress for the fix to #2221. Give me a poke if it looks like I got distracted by a butterfly or something.

![Simpsons GIF of Mr Burns poking Bart](https://media.giphy.com/media/3orif5pvHPCFyBO2Kk/source.gif)

shaunanoordin commented:
## PR Update

This PR is now functionally complete!

- Single Image Viewer now has the following rendering states:
  - No subject resource, _or_ problem loading subject resource, _or_ just initialised => SIV renders nothing (null)
  - Subject resource is loading, _or_ Subject resource has loaded but image file hasn't loaded => SIV renders **placeholder image** (see PNG file in this branch)
  - Subject resource and image file have loaded => SIV renders image file
- ❗ Major change: webpack has been modified to include **file-loader.**
  - Webpack will now load image files (but there's only one, so I didn't think compression is necessary)
- Tests have been fixed! (Edit: have tests been fixed? I need to double check with CI Tests.)

Dev notes:
- The placeholder image is a basic 8kb 800x600px rectangular PNG file to mimic the Invision design.
- This wide placeholder may look a bit out of place on projects with tall Subject images, (i.e. the SIV keeps going placeholder loads, "wide view", image file loads, flash, "tall view") but it won't break anything. 
- I didn't try using an SVG image since I just wanted this in ASAP to fix 2221 so Davy Notebooks can run. Theoretically, this should be possible, but oh man I know I'm going to spend a good day solving for the "perfect" dimension.
- If you want to argue for a default _square_ PNG file... I mean, I was debating that too.

### Status

Aaalllmost ready for review, gotta re-check those tests...

shaunanoordin commented:
Oh wait, butts!

- The placeholder image loads fine on `lib-classifier` ✅ 
- The placeholder image does NOT load fine on `app-project` ❌ 
  - With `yarn dev`, files like `http://local.zooniverse.org:3000/_next/static/development/9fc1a6d27a484c2cc1641ec0ad7fc3f1.png` return a 404
  - This seems like an issue with Next.JS

Thoughts on solutions:
- Spend the next few days figuring out how to solve serving image files on Next.js
- Host the placeholder on zooniverse-static and call it a day

shaunanoordin commented:
I've reverted the addition of the placeholder image PNG into `lib-classifier` since `app-project`'s Next.js doesn't serve image files.

I've also reverted the merge from master because it was borking the tests.

eatyourgreens commented:
NextJS serves static files from the `public` folder. There’s a subject placeholder in there at the moment. It’s used in a couple of places on project pages.

shaunanoordin commented:
Thanks Jim! I'm going to have to look into this.

The current gotcha on my mind is that, if the subject-placeholder.png is in `app-project`'s public folder, then `lib-classifier` won't have a placeholder image on its own (e.g. while running `yarn dev`, or while running unit tests). First things first, I gotta study how app-project uses the placeholder image and see if there's something I can build on.

shaunanoordin commented:
Oh c'mon, why is merging from master breaking all the tests??

shaunanoordin commented:
Note to self: sometimes the tests (after merging from master) just break for no reason whatsoever. Re-run the tests again and sometimes that fixes it.

shaunanoordin commented:
## PR Update

We now have a placeholder image! Big thanks to Cam for helping me get things uploaded. Here's some quick notes from what I've already said [on Slack:](https://zooniverse.slack.com/archives/CANKLB50E/p1623852022083000)

- I've uploaded the Subject Placeholder image for lib-classifier to Zooniverse Static, hooray! https://static.zooniverse.org/www.zooniverse.org/assets/fe-project-subject-placeholder-800x600.png
  - This PR has been updated accordingly 
- Coming back to Jim's suggestion though, I realised that while I was so focused on the impossibility of including app-project's subject placeholder image via code... I overlooked the image file is already uploaded to a URL. https://fe-project.zooniverse.org/subject-placeholder.png
- Question: for future purposes, should all image assets be parked in app-project?
- But: image assets in app-project aren't (currently) sitting behind a CDN, and are served from Next.js (with a cache lifetime of 60s to boot). If we do park all image assets in app-project, do we need to solve for this? (For context, this is a "future proofing" idea, not an urgent topic)

Additional note from Sarah, [on Slack](https://zooniverse.slack.com/archives/CANKLB50E/p1623856875089600):
> I'm interested in trying Next.js v10's Image component which includes several optimizations by default, so if we find it works well, that would be a reason to serve images using the Next.js app. Here's more information about that: https://nextjs.org/docs/basic-features/image-optimization

### Status

In any case, this PR is ready to roll!

srallen reviewed (changes requested):
The stories for the `SingleImageViewer` aren't working because I think the loading state needs to be set to success explicitly now. 

Otherwise, I think this generally looks good for an immediate bug fix. I think for a follow up, we should consider adding a `componentWillUnmount` lifecycle method to the `SingleImageViewerContainer` that calls the `SubjectViewerStore`'s `resetSubject` method, to set the subject back to a loading state since the image `onload` runs every time the page loads. The `resetSubject` method also resets some other parameters, so we would need to test this to make sure this is what we would want to do, however, it doesn't make sense to keep the viewer store in a successfully loading state when navigating away if when navigating back, something needs to load again. Can we track this in a future TODO issue?

shaunanoordin commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SingleImageViewer/SingleImageViewerContainer.js:
```diff
@@ -110,7 +110,12 @@ class SingleImageViewerContainer extends React.Component {
     } = this.props
     const { img } = this.state
-    const { naturalHeight, naturalWidth, src } = img
+    // const { naturalHeight, naturalWidth, src } = img
```
👍

shaunanoordin commented:
Thanks Sarah! Looking into Storybook now

shaunanoordin commented:
## PR Update

- Stories for SingleImageViewer has been fixed.
  - The <SingleImageViewer> component now has explicit `loadingState={success} enableInteractionLayer={false}`
  - I had to explicitly disable the Interaction Layer because otherwise it'll try to render drawing components and break a lot of stuff.
  - This previously wasn't an issue because, while enableInteractionLayer defaults to true, the component wouldn't render the drawing components if loadingState !== success.
- `resetSubject()` enhancement idea tracked in #2251

srallen reviewed (approved):
LGTM. Thanks!
